libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2111,57 +2111,61 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
 		return;
 	}
 
-	auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front()));
-
-	if (!functionPointerType)
+	FunctionType const* externalFunctionType = nullptr;
 	{
-		m_errorReporter.typeError(
-			5511_error,
-			arguments.front()->location(),
-			"Expected first argument to be a function pointer, not \"" +
-			type(*arguments.front())->canonicalName() +
-			"\"."
-		);
-		return;
+		auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front()));
+		if (!functionPointerType)
+		{
+			m_errorReporter.typeError(
+				5511_error,
+				arguments.front()->location(),
+				"Expected first argument to be a function pointer, not \"" +
+				type(*arguments.front())->toString() +
+				"\"."
+			);
+			return;
+		}
+
+		externalFunctionType = functionPointerType->asExternallyCallableFunction(
+													functionPointerType->kind() == FunctionType::Kind::DelegateCall
```

Comment:
I felt this made it much more readable.
Plus there is a check in the code below, that carries an assert that it can't be DelegateCall. 
So, having an assert here as well seemed repetitive.